aws-iot-device-sdk/index.d.ts

```diff
@@ -0,0 +1,402 @@
+   * the clientToken will be supplied as one of the parameters, allowing the
+   * application to keep track of the status of each operation. The caller may
+   * create their own clientToken value; if stateObject contains a clientToken
+   * property, that will be used rather than the internally generated value.
+   * Note that it should be of atomic type (i.e. numeric or string).
+   * This function returns "null" if an operation is already in progress.
+   */
+  update(thingName: string, stateObject: any): string | null;
+
+  /**
+   * Get the current state of the Thing Shadow named thingName, which must
+   * have been previously registered using awsIot.thingShadow#register().
+   * The thingShadow class will subscribe to all applicable topics and
+   * publish on the get sub-topic.
+   *
+   * This function returns a clientToken, which is a unique value
+   * associated with the get operation. When a "status or "timeout" event
+   * is emitted, the clientToken will be supplied as one of the parameters,
+   * allowing the application to keep track of the status of each operation.
+   * The caller may supply their own clientToken value (optional); if
+   * supplied, the value of clientToken will be used rather than the
+   * internally generated value. Note that this value should be of atomic
+   * type (i.e. numeric or string). This function returns "null" if an
+   * operation is already in progress.
+   */
+  get(thingName: string, clientToken?: string): string | null;
+
+  /**
+   * Delete the Thing Shadow named thingName, which must have been previously
+   * registered using awsIot.thingShadow#register(). The thingShadow class
+   * will subscribe to all applicable topics and publish on the delete sub-topic.
+   *
+   * This function returns a clientToken, which is a unique value associated
+   * with the delete operation. When a "status" or "timeout" event is
+   * emitted, the clientToken will be supplied as one of the parameters,
+   * allowing the application to keep track of the status of each operation.
+   * The caller may supply their own clientToken value (optional); if
+   * supplied, the value of clientToken will be used rather than the
+   * internally generated value. Note that this value should be of atomic
+   * type (i.e. numeric or string). This function returns "null" if an
+   * operation is already in progress.
+   */
+  delete(thingName: string, clientToken?: string): string | null;
+
+  // The following publish, subscribe, unsubscribe and end Definitions
+  // are copied from the mqtt definition as they are re-surfaced through
+  // thingShadow
+  // https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e7772769fab8c206449ce9673cac417370330aa9/mqtt/mqtt.d.ts#L199
+
+  /**
+   * Publish a message to a topic
+   *
+   * @param topic
+   * @param message
+   * @param options
+   * @param callback
+   */
+  publish(topic: string, message: Buffer, options?: mqtt.ClientPublishOptions, callback?: Function): mqtt.Client;
```

Comment:
So these were all copied verbatim from the [mqtt definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3abf7a7bd48328caa63a30c0b26467f8c17bff12/mqtt/mqtt.d.ts#L191-L200) since aws-iot-device-sdk marshals them straight through. I initially only wanted to make sure that they match exactly but I could certainly figure out what that callback is and then contribute it to mqtt after. I'll give it a shot